File: 240129/트로미노/tromino.py
R, C = map(int, input().split())

board = []
for _ in range(R):
    row = list(map(int, input().split()))
    board.append(row)

visited = [[False for _ in range(C)] for _ in range(R)]
# dfs 3개로 해서 풀기

dr = [-1, 0, 1, 0]
dc = [0, -1, 0, 1]

answer = 0

def dfs(r, c, s, l):
    global answer
    if l == 3:
        # print("s", s)
        answer = max(answer, sum(s))
        return

    for i in range(4):
        rr = r + dr[i]
        cc = c + dc[i]

        if 0<=rr<R and 0<=cc<C and not visited[rr][cc]:
            visited[rr][cc] = True
            dfs(rr, cc, s+[board[rr][cc]], l+1)
            visited[rr][cc] = False
    

for r in range(R):
    for c in range(C):
        visited[r][c] = True
        dfs(r, c, [board[r][c]], 1)
        visited[r][c] = False
print(answer)

File: 240404/트로미노/tromino.swift
let NM = readLine()!.split(separator: " ").map { Int(String($0))! }
let N = NM[0]
let M = NM[1]

var board: [[Int]] = []
for _ in 0..<N {
    let row = readLine()!.split(separator: " ").map { Int(String($0))! }
    board.append(row)
}

// print(board)

var answer: [Int] = []
var dr = [-1, 0, 1, 0]
var dc = [0, -1, 0, 1]

func dfs(r: Int, c: Int, depth: Int, visited: [[Int]]) {
    if depth == 3 {
        answer.append(visited[r][c])
        return
    }

    var newVisited = visited

    for i in 0..<4 {
        let rr = r + dr[i]
        let cc = c + dc[i]

        if rr >= 0 && rr < N && cc >= 0 && cc < M && newVisited[rr][cc] == 0 {
            newVisited[rr][cc] = newVisited[r][c] + board[rr][cc]
            dfs(r: rr, c: cc, depth: depth+1, visited: newVisited)
        }
    }

}

for r in 0..<N {
    for c in 0..<M {
        var visited = Array(repeating: Array(repeating: 0, count: M), count: N)
        visited[r][c] = board[r][c]
        dfs(r: r, c: c, depth: 1, visited: visited)
    }
}
// print(answer)
answer.sort()
print(answer.last!)